From a8e595a9635426c318e395aa2989e72b77996183 Mon Sep 17 00:00:00 2001
From: Guobing Chen <guobing.chen@intel.com>
Date: Tue, 4 Jun 2019 16:06:28 +0800
Subject: [PATCH] ported blas-ht patch

This is an updated patch to port Arjan's blas-ht patch to adapt to
openBLAS v0.3.6

Signed-off-by: Guobing Chen <guobing.chen@intel.com>
---
 driver/others/blas_server.c | 18 +++++++++---------
 driver/others/memory.c      | 24 ++++++++++++++++++++++++
 2 files changed, 33 insertions(+), 9 deletions(-)

diff --git a/driver/others/memory.c b/driver/others/memory.c
index f67cb01f..9327917e 100644
--- a/driver/others/memory.c
+++ b/driver/others/memory.c
@@ -1737,6 +1737,22 @@ void goto_set_num_threads(int num_threads) {};
 #else
 
 #if defined(OS_LINUX) || defined(OS_SUNOS)
+static hyperthreading_divider(void)
+{
+       FILE *file;
+       char buffer[16];
+       file = fopen("/sys/devices/system/cpu/cpu0/topology/thread_siblings_list", "r");
+       if (!file)
+               return 1;
+       buffer[0] = 0;
+       buffer[1] = 0;
+       fread(buffer, 1, 3, file);
+       fclose(file);
+       if (buffer[1] == ',')
+               return 2;
+       return 1;
+}
+
 #ifndef NO_AFFINITY
 int get_num_procs(void);
 #else
@@ -1761,6 +1777,10 @@ int get_num_procs(void) {
   return nums;
 #endif
 
+#if defined(linux)
+   nums = nums / hyperthreading_divider();
+#endif
+
 #if !defined(__GLIBC_PREREQ)
   return nums;
 #else
@@ -1794,6 +1814,8 @@ int get_num_procs(void) {
     }
     ret = CPU_COUNT_S(size,cpusetp);
     if (ret > 0 && ret < nums) nums = ret;	
+    if (nums == sysconf(_SC_NPROCESSORS_CONF)) /* no mask, fall back to default HT policy */
+        nums = nums / hyperthreading_divider();
     CPU_FREE(cpusetp);
     return nums;
   } else {
@@ -1803,6 +1825,8 @@ int get_num_procs(void) {
     }
     ret = CPU_COUNT(&cpuset);
     if (ret > 0 && ret < nums) nums = ret;	
+    if (nums == sysconf(_SC_NPROCESSORS_CONF)) /* no mask, fall back to default HT policy */
+        nums = nums / hyperthreading_divider();
     return nums;
   }
  #endif
-- 
2.21.0

